Classification/FeatureSelection/src/main/java/org/tribuo/classification/fs/FS_Wrapper_Approaches/Optimizers/CuckooSearchOptimizer.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    /**
+     * @param transferFunction The transfer function to convert continuous values to binary ones
+     * @param populationSize The size of the solution in the initial population
+     * @param stepSizeScaling The cuckoo step size
+     * @param lambda The lambda of the levy flight function
+     * @param worstNestProbability The fraction of the nests to be abandoned
+     * @param delta The delta that is used in the abandon nest function
+     * @param maxIteration The number of times that is used to enhance generation
+     */
+    public CuckooSearchOptimizer(TransferFunction transferFunction, int populationSize, double stepSizeScaling, double lambda, double worstNestProbability, double delta, int maxIteration) {
+        this.transferFunction = transferFunction;
+        this.populationSize = populationSize;
+        this.stepSizeScaling = stepSizeScaling;
+        this.lambda = lambda;
+        this.worstNestProbability = worstNestProbability;
+        this.delta = delta;
+        this.maxIteration = maxIteration;
+    }
+
+    /**
+     * @param totalNumberOfFeatures The number of features in the given dataset
+     * @return The population of subsets of selected features
+     */
+    private int[][] GeneratePopulation(int totalNumberOfFeatures) {
+        setOfSolutions = new int[this.populationSize][totalNumberOfFeatures];
+        for (int[] subSet : setOfSolutions)
+            System.arraycopy(new Random().ints(totalNumberOfFeatures, 0, 2).toArray(), 0, subSet, 0, setOfSolutions[0].length);
+        return setOfSolutions;
+    }
+
+    /**
+     * Does this feature selection algorithm return an ordered feature set?
+     *
+     * @return True if the set is ordered.
+     */
+    @Override
+    public boolean isOrdered() {
+        return true;
+    }
+
+    /**
+     * Selects features according to this selection algorithm from the specified dataset.
+     * @param dataset The dataset to use.
+     * @return A selected feature set.
+     */
+    @Override
+    public SelectedFeatureSet select(Dataset<Label> dataset) {
+        ImmutableFeatureMap FMap = new ImmutableFeatureMap(dataset.getFeatureMap());
+        setOfSolutions = GeneratePopulation(dataset.getFeatureMap().size());
+        List<FeatureSet_FScore_Container> subSet_fScores = new ArrayList<>();
+        SelectedFeatureSet selectedFeatureSet = null;
+        // Update the solution based on the levy flight function
+        for (int i = 0; i < maxIteration; i++) {
+            IntStream.range(0, setOfSolutions.length).parallel().forEach(subSet -> {
+                AtomicInteger currentIter = new AtomicInteger(subSet);
+                int[] evolvedSolution = Arrays.stream(setOfSolutions[subSet]).map(x -> Binarizing.discreteValue(transferFunction, x + stepSizeScaling * Math.pow(currentIter.get() + 1, -lambda))).toArray();
+                int[] randomCuckoo = setOfSolutions[new Random().nextInt(setOfSolutions.length)];

```

Comment:
`new Random()` is not allowed in Tribuo, it introduces non-determinism and we can't capture accurate provenance information. All the uses of an RNG need to be refactored so there's a single instance of `SplittableRandom` in the class, which is then split in a synchronized block into a local RNG for the `select` method, then provenance is captured, the invocation count is incremented (which will need explicit capturing in the provenance too), and then all RNGs are split from the local one in a deterministic way. You can see the right idiom in `KMeansTrainer`.